mosec/dry_run.py

```diff
@@ -0,0 +1,151 @@
+    :py:meth:`append_worker <mosec.server.Server.append_worker>`, the 1st one
+    will be used.
+    """
+
+    def __init__(
+        self,
+        workers: List[type[Worker]],
+        batches: List[int],
+        envs: List[None | List[Dict[str, str]]],
+    ):
+        """Init dry runner."""
+        logger.info("init dry runner for %s", workers)
+        self.process_context = SpawnContext()
+        self.workers = workers
+        self.batches = batches
+        self.envs = envs
+        self.pool: List[SpawnProcess] = []
+        self.sender_pipes: List[PipeConnection] = []
+        self.receiver_pipes: List[PipeConnection] = []
+
+    def new_pipe(self):
+        """Create new pipe for dry run workers to communicate."""
+        receiver, sender = self.process_context.Pipe(duplex=False)
+        self.sender_pipes.append(sender)
+        self.receiver_pipes.append(receiver)
+
+    def run(self):
+        """Execute thr dry run process."""
+        self.new_pipe()
+        for i, (worker, batch, env) in enumerate(
+            zip(self.workers, self.batches, self.envs)
+        ):
+            self.new_pipe()
+            coordinator = self.process_context.Process(
+                target=dry_run_func,
+                args=(
+                    worker,
+                    batch,
+                    self.receiver_pipes[-2],
+                    self.sender_pipes[-1],
+                    i == 0,
+                ),
+                daemon=True,
+            )
+
+            with env_var_context(env, 0):
+                coordinator.start()
+
+            self.pool.append(coordinator)
+
+        logger.info("dry run init successful")
+        self.warmup()
+        logger.info("dry run exit")
+
+    def warmup(self):
+        """Warmup the service.
+
+        Only one example will be used.
+        """
+        ingress = self.workers[0]
+        example = None
+        if ingress.example:
+            assert isinstance(
+                ingress.example, list
+            ), "`example` should be a list of data"
+            example = ingress.example[0]
+        elif ingress.multi_examples:
+            assert isinstance(ingress.multi_examples, list) and isinstance(
+                ingress.multi_examples[0], list
+            ), "`multi_examples` should be a list of list data"
+            example = ingress.multi_examples[0][0]
+
+        if not example:
+            logger.info("cannot find the example in the 1st stage worker, skip warmup")
+            return
+
+        sender, receiver = self.sender_pipes[0], self.receiver_pipes[-1]
+        start_time = time.perf_counter()
+        sender.send(example)
+        res = receiver.recv_bytes()
+        duration = time.perf_counter() - start_time
+        logger.info(
+            "dry run result: %s",
+            json.dumps(
+                {
+                    "req": example,
+                    "size": len(res),
```

Comment:
Shall we specify this more clearly: input data size, output data size for each stage, and the duration of each stage? Also maybe we need do the warmup first, then measure the timing of the second forward pass, since downloading and warming up takes a lot of time .

Reply:
I don't know what should be done during the dry run.

Ideally, we should collect enough metrics to propose suitable online serving arguments.

Maybe we can do it in another PR?

Reply:
Definitely collecting metrics can be useful, yep we can add them later